### PR TITLE
Noisy student training for wenet

### PR DESCRIPTION
**Pull request by NevermoreCY**
Here, we provide a recipe to run Noisy Student Training (NST) with LM filter strategy using AISHELL-1 as supervised data and WenetSpeech as unsupervised data from [our paper](https://arxiv.org/abs/2211.04717).

The example codes are stored under examples/aishell/NST with a detailed guideline and results in readme.md. we mainly modified the script for "run.sh" and added "executor_nst.py" ,"train_nst.py" as well as some auxiliary codes and examples under local directory.

### CODE REVIEW

**Comment by wd929** [comment]
@robin1001 hi Binbin, would you mind reviewing this PR? Thanks!!

File: wenet/bin/train_nst.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+    train_supervised_conf = configs['supervised_dataset_conf']
+    train_unsupervised_conf = configs['unsupervised_dataset_conf']
+
+    print("*********train setting**************")
+    print("supervised data list = ", args.train_data_supervised)
+    print("unsupervised data list = ", args.train_data_unsupervised)
+
+    cv_conf = copy.deepcopy(train_supervised_conf)
+    # cv_conf['speed_perturb'] = False
```
Comment (robin1001) [inline wenet/bin/train_nst.py:151]: why use speed_perturb and spec_aug in cv?

Comment (NevermoreCY) [reply wenet/bin/train_nst.py]: Hi robin, thanks for pointing out this issue, the speed_perturb, spech_aug as well as shuffle should be set to false for CV.  I traced back my history and found it was a mis-comment during my recent review. In our experiment cluster, it was set to be False. 
eae283e8e3a124bd12bca4d8f53cd6c

**Comment by NevermoreCY** [comment]
The flake8-bugbear is updated to version 22.12.6 from 22.10.27 and they modified the test for zip function.
 It seems codes contains zip() from other section didn't pass the flake8 test.

**Comment by xingchensong** [comment]
> The flake8-bugbear is updated to version 22.12.6 from 22.10.27 and they modified the test for zip function. It seems codes contains zip() from other section didn't pass the flake8 test.

fixed, https://github.com/wenet-e2e/wenet/pull/1604, plz rebase your commits

**Comment by NevermoreCY** [comment]
> 
It works now, thanks!

File: wenet/bin/train_nst.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+import logging
+import os
+import torch
+import torch.distributed as dist
+import torch.optim as optim
+import yaml
+from tensorboardX import SummaryWriter
+from torch.utils.data import DataLoader
+from wenet.dataset.dataset import Dataset
+from wenet.transformer.asr_model import init_asr_model
```
Comment (xingchensong) [inline wenet/bin/train_nst.py:27]: What's your base commits? we have refactor our model initialization in this PR https://github.com/wenet-e2e/wenet/pull/1280, call `wenet.utils.init_model.init_model` instead of `wenet.transformer.asr_model.init_asr_model`

File: wenet/bin/train_nst.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    train_unsupervised_conf = configs['unsupervised_dataset_conf']
+
+    print("*********train setting**************")
+    print("supervised data list = ", args.train_data_supervised)
+    print("unsupervised data list = ", args.train_data_unsupervised)
+
+    cv_conf = copy.deepcopy(train_supervised_conf)
+    cv_conf['speed_perturb'] = False
+    cv_conf['spec_aug'] = False
+    cv_conf['shuffle'] = False
```
Comment (xingchensong) [inline wenet/bin/train_nst.py:153]: better to add:

```py
cv_conf['spec_sub'] = False
cv_conf['spec_trim'] = False
```

File: wenet/utils/executor_nst.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                    logging.info(f"{name} reaches end of "
+                                 f"dataloader for rank {rank}")
+                    break
+
+                batch_idx += 1
+                feats = feats.to(device)
+                target = target.to(device)
+                feats_lengths = feats_lengths.to(device)
+                target_lengths = target_lengths.to(device)
+                # print("Time after to device ", time.time())
```
Comment (xingchensong) [inline wenet/utils/executor_nst.py:70]: I think this line can be removed.

File: wenet/utils/executor_nst.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                key, feats, target, feats_lengths, target_lengths = batch
+                feats = feats.to(device)
+                target = target.to(device)
+                feats_lengths = feats_lengths.to(device)
+                target_lengths = target_lengths.to(device)
+                num_utts = target_lengths.size(0)
+                if num_utts == 0:
+                    continue
+                loss, loss_att, loss_ctc = model(feats, feats_lengths, target,
+                                                 target_lengths)
```
Comment (xingchensong) [inline wenet/utils/executor_nst.py:154]: We have changed the return type from multiple tensors to loss_dict, plz see https://github.com/wenet-e2e/wenet/pull/1261/files#r909610210

File: wenet/utils/executor_nst.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                if torch.isfinite(loss):
+                    num_seen_utts += num_utts
+                    total_loss += loss.item() * num_utts
+                if batch_idx % log_interval == 0:
+                    log_str = 'CV Batch {}/{} loss {:.6f} '.format(
+                        epoch, batch_idx, loss.item())
+                    if loss_att is not None:
+                        log_str += 'loss_att {:.6f} '.format(loss_att.item())
+                    if loss_ctc is not None:
+                        log_str += 'loss_ctc {:.6f} '.format(loss_ctc.item())
```
Comment (xingchensong) [inline wenet/utils/executor_nst.py:164]: same here https://github.com/wenet-e2e/wenet/pull/1261/files#diff-317ef7c1c126ff1e43242e52cc727d6979719019544e7e901173788fa2435c16R147-R150

File: examples/aishell/NST/run_nst.sh [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+
+
+# This is an augmented version of aishell-1 "run.sh" to make the code compatible with noisy student training
+
+. ./path.sh || exit 1;
+
+# Use this to control how many gpu you use, It's 1-gpu training if you specify
+# just 1gpu, otherwise it's is multiple gpu training based on DDP in pytorch
+export CUDA_VISIBLE_DEVICES="0,1,2,3,4,5,6,7"
+#export CUDA_VISIBLE_DEVICES="0"
```
Comment (xingchensong) [inline examples/aishell/NST/run_nst.sh:26]: we can remove this line

File: examples/aishell/NST/run_nst.sh [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+
+train_set=train
+
+train_config=conf/train_conformer_nst.yaml
+
+cmvn=true
+#dir=exp/conformer_wenet1k_nst5_LM_diff_leq_10
+#checkpoint=exp/conformer_wenet1k_nst0_sr3_v2/avg_30.pt
+
+#avg_30.pt
```
Comment (xingchensong) [inline examples/aishell/NST/run_nst.sh:92]: same here

File: examples/aishell/NST/run_nst.sh [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+
+# use average_checkpoint will get better result
+# ??
+average_checkpoint=true
+target_pt=80
+decode_checkpoint=$dir/$target_pt.pt
+
+# here we only use attention_rescoring for NST
+decode_modes="attention_rescoring"
+#decode_modes="ctc_greedy_search ctc_prefix_beam_search attention attention_rescoring"
```
Comment (xingchensong) [inline examples/aishell/NST/run_nst.sh:102]: only keep one of therm

File: examples/aishell/NST/run_nst.sh [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+export NCCL_DEBUG=INFO
+stage=1 # start from 0 if you need to start from data preparation
+stop_stage=8
+
+# here are extra parameters used in NST
+cer_out_dir=""
+dir=""
+pseudo_data_list=""
+supervised_data_list=""
+gcmvn=""
```
Comment (xingchensong) [inline examples/aishell/NST/run_nst.sh:40]: seems that `gcmvn` is never used.

File: examples/aishell/NST/run_nst.sh [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+# if you have label for unsupervised dataset, set label = 1 other wise keep it 0
+# For each gpu or cpu, you can run with different job_num to perform data-wise parallel computing.
+if [ ${stage} -le 5 ] && [ ${stop_stage} -ge 5 ] && [ ${enable_nst} -eq 0 ]; then
+  echo "********stage 5 start time : $now ********"
+  python get_wav_labels.py \
+    --dir_split data/train/${dir_split} \
+    --hypo_name /$hypo_name \
+    --wav_dir $wav_dir\
+    --job_num $job_num \
+    --label $label
```
Comment (xingchensong) [inline examples/aishell/NST/run_nst.sh:343]: `PYTHONPATH` in `path.sh` does not include $PWD/local, I this line 338 should be
```py
python local/get_wav_labels.py
```

File: examples/aishell/NST/README.md [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+* Training info: lr 0.002, batch size 32, 8 gpu, acc_grad 4, 240 epochs, dither 0.1
+* Decoding info: ctc_weight 0.3, average_num 30
+
+
+| Supervised               | Unsupervised | Test CER |
+|--------------------------|--------------|----------|
+| AISHELL-1 Only           | ----         | 4.85     |
+| AISHELL-1+WenetSpeech    | ----         | 3.54     |
+| AISHELL-1+AISHELL-2      | ----         | 1.01     |
+| AISHELL-1 (standard NST) | WenetSpeech  | 5.52     |
```
Comment (xingchensong) [inline examples/aishell/NST/README.md:93]: According to your .yaml configuration, this is a **non-streaming model** and  `4.85` is obtained from **attention restoring**. 

I do think we'd better mention those Infos directly in README.

**Comment by wd929** [comment]
Sure. Will do!

**Comment by wd929** [comment]
Could you please review this PR? @robin1001 and @xingchensong Thanks!!^^

File: examples/aishell/NST/data_example/train/data_aishell.list [hunk 13]
```diff
@@ -0,0 +1,11 @@
+data/train/shards/shards_000000000.tar
```
Comment (robin1001) [inline examples/aishell/NST/data_example/train/data_aishell.list:1]: please remove data_example/train dir

File: examples/aishell/NST/data_example/train/wav_dir/ID001.txt [hunk 14]
```diff
@@ -0,0 +1 @@
+你好你好
```
Comment (robin1001) [inline examples/aishell/NST/data_example/train/wav_dir/ID001.txt:1]: remove wav_dir

File: examples/aishell/NST/data_example/train/wenet_1khr.list [hunk 15]
```diff
@@ -0,0 +1,2 @@
+data/train/wenet_1khr_tar//dir0_000000.tar
+data/train/wenet_1khr_tar//dir0_000001.tar
```
Comment (robin1001) [inline examples/aishell/NST/data_example/train/wenet_1khr.list:2]: remove the file

File: examples/aishell/NST/run_nst.sh [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+  done
+  wait
+fi
+
+
+# split the (unsupervised) datalist into N sublists, where N depends on the number of available cpu in your cluster.
+# when making inference, we compute N sublist in parallel.
+if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ] && [ ${enable_nst} -eq 0 ]; then
+  echo "********step 3 start time : $now ********"
+  python split_data_list.py \
```
Comment (xingchensong) [inline examples/aishell/NST/run_nst.sh:253]: `python local/split_data_list.py`

**Comment by xingchensong** [comment]
Great Job, LGTM.

**Comment by NevermoreCY** [comment]
> please remove the files which are not required for the recipe.

The data example directory, train_nst.py and executor.py has been removed, the problems you mentioned should be sovled.

**Review by robin1001 — APPROVED** [review]
